File: solutions/3517-shortest-distance-after-road-addition-queries-i/shortest-distance-after-road-addition-queries-i.swift
class Solution {
    func shortestDistanceAfterQueries(_ n: Int, _ queries: [[Int]]) -> [Int] {
        var adj = Array(repeating: [Int](), count: n)
        for i in 0..<n {
            adj[i] = [i + 1]
        }
        
        func shortestPath() -> Int {
            var queue = [(0, 0)] // node, length
            var visit = Set<Int>()
            visit.insert(0)
            var queueIndex = 0
            
            while queueIndex < queue.count {
                let (cur, length) = queue[queueIndex]
                queueIndex += 1
                
                if cur == n - 1 {
                    return length
                }
                
                for nei in adj[cur] {
                    if !visit.contains(nei) {
                        queue.append((nei, length + 1))
                        visit.insert(nei)
                    }
                }
            }
            return -1
        }
        
        var res = [Int]()
        for query in queries {
            let src = query[0], dst = query[1]
            adj[src].append(dst)
            res.append(shortestPath())
        }
        return res
    }
}

File: solutions/3517-shortest-distance-after-road-addition-queries-i/shortest-distance-after-road-addition-queries-i.ts
function shortestDistanceAfterQueries(n: number, queries: number[][]): number[] {
    const adj: number[][] = Array(n).fill(0).map((_, i) => [i + 1]);
    
    const shortestPath = (): number => {
        const q: [number, number][] = [[0, 0]]; // node, length
        const visit = new Set<number>();
        visit.add(0);
        
        while (q.length > 0) {
            const [cur, length] = q.shift()!;
            
            if (cur === n - 1) {
                return length;
            }
            
            for (const nei of adj[cur]) {
                if (!visit.has(nei)) {
                    q.push([nei, length + 1]);
                    visit.add(nei);
                }
            }
        }
        return -1;
    };
    
    const res: number[] = [];
    for (const [src, dst] of queries) {
        adj[src].push(dst);
        res.push(shortestPath());
    }
    return res;
}